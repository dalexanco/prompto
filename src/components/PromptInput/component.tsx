import React from "react";
import MagnifyingGlassIcon from "../../icons/magnifying-glass";

export function PromptInput(): JSX.Element {
    return (
        <div className="inline-flex items-center bg-white px-2 py-1">
            <MagnifyingGlassIcon className="w-4 h-4 stroke-gray-500" />
            <input
                autoFocus
                placeholder="Search or command..."
                type="text"
                className="text-sm whitespace-nowrap px-2 my-1 outline-none placeholder:italic"
            />
        </div>
    );
}
