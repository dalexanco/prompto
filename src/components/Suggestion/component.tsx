import React from "react";
import classNames from "classnames";

import MagnifyingGlassIcon from "../../icons/magnifying-glass";
import { PromptCommandType } from "../../types/commands";

interface SuggestionProps {
    title: string;
    description?: string;
    type: PromptCommandType;
    hasFocus?: boolean;
}

export function Suggestion({
    title,
    description,
    hasFocus = false,
}: SuggestionProps): JSX.Element {
    const wrapperClass = classNames("flex flex-row items-start bg-white", {
        ["bg-slate-50"]: hasFocus,
    });
    const iconClass = classNames(
        "flex self-center rounded-md p-2 m-2 bg-gray-200",
        {
            ["bg-gray-200"]: hasFocus,
            ["bg-gray-100"]: !hasFocus,
        },
    );
    return (
        <div className={wrapperClass}>
            <div className={iconClass}>
                <MagnifyingGlassIcon className="w-4 h-4 stroke-gray-500" />
            </div>
            <div className="mx-2 self-center min-w-0">
                <p className="text-sm font-medium overflow-ellipsis whitespace-nowrap overflow-hidden">
                    {title}
                </p>
                {hasFocus && description && (
                    <p className="text-xs text-gray-500 overflow-ellipsis whitespace-nowrap overflow-hidden">
                        {description}
                    </p>
                )}
            </div>
        </div>
    );
}
