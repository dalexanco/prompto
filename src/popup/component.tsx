import React, { FormEvent, useCallback, useState } from "react";
import { useKeyPressEvent } from "react-use";
import browser from "webextension-polyfill";

import { PromptInput } from "../components/PromptInput";
import { Suggestion } from "../components/Suggestion";
import useBookmarks from "../hooks/use-bookmarks";
import ChevronDown from "../icons/chevron-down";
import ChevronUp from "../icons/chevron-up";
import { PromptCommand, PromptCommandType } from "../types/commands";
import css from "./styles.module.css";
import useExistingTabs from "@src/hooks/use-existing-tabs";
import useExecute from "@src/hooks/use-execute";

const useSuggestionFocus = (
    suggestions: PromptCommand[],
): PromptCommand | undefined => {
    const [focusedLine, setFocus] = useState(0);
    const moveFocus = useCallback(
        (value) => {
            setFocus(
                (suggestions.length + focusedLine + value) % suggestions.length,
            );
        },
        [focusedLine, setFocus, suggestions.length],
    );
    useKeyPressEvent("ArrowUp", () => moveFocus(-1));
    useKeyPressEvent("ArrowDown", () => moveFocus(1));

    return suggestions[focusedLine];
};

export function Popup(): JSX.Element {
    React.useEffect(() => {
        browser.runtime.sendMessage({ popupMounted: true });
    }, []);

    const [inputValue, updateInput] = useState("");
    const { results: suggestionsBookmarks } = useBookmarks(inputValue);
    const { results: suggestionsTabs } = useExistingTabs(inputValue);
    const suggestions = [
        ...suggestionsTabs,
        ...suggestionsBookmarks,
    ] as PromptCommand[];
    const focusSuggestion = useSuggestionFocus(suggestions);

    // Form validation
    const onSubmit = (event: FormEvent) => {
        event.preventDefault();
        useExecute(focusSuggestion);
    };

    return (
        <div className={css.popupContainer}>
            <form className="border-b border-b-gray-200" onSubmit={onSubmit}>
                <PromptInput onChange={updateInput} />
            </form>
            <section className="flex flex-grow flex-col items-stretch">
                {suggestions.map((suggestion, index) => (
                    <Suggestion
                        suggestion={suggestion}
                        key={suggestion.key}
                        hasFocus={suggestion.key === focusSuggestion?.key}
                    />
                ))}
            </section>
            <footer className="flex justify-end border-t border-t-gray-200 px-2 py-1 bg-gray-50">
                <p className="justify-self-end text-xs text-gray-500 italic">
                    Use{" "}
                    <span className=" inline-flex bg-gray-200 px-1 py-1 rounded-sm">
                        <ChevronUp className="w-2 h-2 inline-flex" />
                    </span>{" "}
                    <span className=" inline-flex bg-gray-200 px-1 py-1 rounded-sm">
                        <ChevronDown className="w-2 h-2 inline-flex" />
                    </span>{" "}
                    to navigate
                </p>
            </footer>
        </div>
    );
}
