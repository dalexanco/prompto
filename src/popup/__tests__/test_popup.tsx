import * as React from "react";
import { Popup } from "../component";
import render from "react-test-renderer";

it("component renders", () => {
    //const tree = renderer.create(<Popup />).toJSON();
    //expect(tree).toMatchSnapshot();
});
