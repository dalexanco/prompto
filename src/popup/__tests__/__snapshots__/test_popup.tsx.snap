// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders 1`] = `
<div
  className="popupContainer"
>
  <form
    className="border-b border-b-gray-200"
    onSubmit={[Function]}
  >
    <div
      className="inline-flex items-center bg-white px-4 py-2"
    >
      <svg
        className="w-4 h-4 stroke-gray-500"
        fill="none"
        strokeWidth={2}
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
          strokeLinecap="round"
          strokeLinejoin="round"
        />
      </svg>
      <input
        autoFocus={true}
        className="text-base whitespace-nowrap px-2 my-1 ml-4 outline-none placeholder:italic"
        onChange={[Function]}
        placeholder="Search or command..."
        type="text"
      />
    </div>
  </form>
  <ul
    className="flex flex-grow flex-col items-stretch overflow-y-scroll"
  />
  <footer
    className="flex justify-end border-t border-t-gray-200 p-2 bg-gray-50"
  >
    <p
      className="justify-self-end text-xs text-gray-500 italic"
    >
      Use
       
      <span
        className=" inline-flex bg-gray-200 px-1 py-1 rounded-sm"
      >
        <svg
          className="w-2 h-2 inline-flex"
          fill="none"
          stroke="currentColor"
          strokeWidth={3}
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.5 15.75l7.5-7.5 7.5 7.5"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </svg>
      </span>
       
      <span
        className=" inline-flex bg-gray-200 px-1 py-1 rounded-sm"
      >
        <svg
          className="w-2 h-2 inline-flex"
          fill="none"
          stroke="currentColor"
          strokeWidth={3}
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19.5 8.25l-7.5 7.5-7.5-7.5"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </svg>
      </span>
       
      to navigate
    </p>
  </footer>
</div>
`;
